### PR TITLE
Build docker image as part of agent pipeline

### PR DESCRIPTION
**Pull request by lox**
Currently docker images are build via an async trigger to a pipeline that builds from https://github.com/buildkite/docker-buildkite-agent.

That repository builds images for all our versions (beta, stable and edge), regardless of what was actually built. 

This moves the docker packaging and building into the agent pipeline and only builds images for the current agent version. The following tags are created:

* For beta releases: 3.0-beta.42.2243 is tagged as beta, 3.0-beta.42, 3.0 and 3
* For edge releases: 3.0-beta.42.2243 is tagged as edge and edge-build-2243
* For stable release: 2.6.10 is tagged as stable, latest, 2.6.10, 2.6 and 2

Once this is in place, we can probably entirely remove the old https://github.com/buildkite/docker-buildkite-agent repo.

### CODE REVIEW

**Comment by lox** [comment]
@toolmantim I'll leave this one for a while in case you get a chance to eye ball it and provide feels. We can probably ship 3.0.0 without this change.

File: scripts/publish-docker-images.sh [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+# variants of beta/experimental - e.g 3.0-beta.16
+if [[ "$CODENAME" == "experimental" ]] ; then
+  for tag in beta $(parse_version "$AGENT_VERSION") ; do
+    if is_stable_version "$tag" ; then
+      echo "--- :docker: Skipping tagging stable ${DOCKER_IMAGE}:${tag}"
+      continue
+    fi
+    release_image "$tag"
+  done
+fi
```
Comment (sj26) [inline scripts/publish-docker-images.sh]: I thought beta == unstable, and edge == experimental?

**Comment by sj26** [comment]
This is great, I wanted to do this while looking at agent packaging recently. 💯

**Comment by toolmantim** [comment]
Don’t wait for me!

**Comment by lox** [comment]
Such merge anxiety. 😱

File: scripts/publish-docker-images.sh (hunk 1 above)
Comment (lox) [reply scripts/publish-docker-images.sh]: Shoot, you are right. Good catch!

**Comment by toolmantim** [comment]
What’s on the merge anxiety top list? Perhaps there’s some tests we can do to help.

**Comment by lox** [comment]
Nah, more just I was nervous about creating more technical debt. This is the right way to go though, let's ship it.

File: scripts/publish-docker-images.sh [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+if [[ "$DOCKER_PULL" =~ (true|1) ]] ; then
+  echo '--- :docker: Pulling prebuilt image'
+  docker pull "$PREBUILT_IMAGE"
+fi
+
+# variants of edge/experimental
+if [[ "$CODENAME" == "experimental" ]] ; then
+  release_image "edge-build-${BUILDKITE_BUILD_NUMBER}"
+  release_image "edge"
```
Comment (sj26) [inline scripts/publish-docker-images.sh]: I don't think we want to overwrite the "edge" tag with a new build on the 2-6-stable branch. Can we limit this somehow?

Comment (sj26) [reply scripts/publish-docker-images.sh]: Probably something like the is_stable_version protection below?

Comment (lox) [reply scripts/publish-docker-images.sh]: But this code won't ever be on that branch, will it?

Comment (lox) [reply scripts/publish-docker-images.sh]: But, I will anyway.

Comment (sj26) [reply scripts/publish-docker-images.sh]: @lox good point!

**Pull request merged**